readme.md
[![build status](https://secure.travis-ci.org/phuesler/ain.png)](http://travis-ci.org/phuesler/ain)
# ain*


Brain-free [syslog](http://en.wikipedia.org/wiki/Syslog)** logging for 
[node.js](http://nodejs.org).

*Ain* is written with full compatibility with *node.js* `console` module. It 
implements all `console` functions and formatting. Also *ain* supports UTF-8 
(tested on Debian Testing/Sid).

*Ain* can send messages by UDP to `127.0.0.1:514` or to the a unix socket; 
/dev/log on Linux and /var/run/syslog on Mac OS X. The unix sockets only
work for the 0.4.x versions of node.js, unix_dgram sockets support has
been [removed](http://groups.google.com/group/nodejs/browse_thread/thread/882ce172ec463f52/62e392bb0f32a7cb) from > 0.5.x.

*In the Phoenician alphabet letter "ain" indicates eye.

**All examples tested under Ubuntu `rsyslog`. On other operating 
systems and logging daemons settings and paths may differ.

## Installation

You can install *ain* as usual - by copy the "ain" directory in your 
`~/.node_modules` or via *npm*

    npm install ain2

## Usage

Usage of *ain* is very similar to the *node.js* console. The following example 
demonstrates the replacement of the console:

    var SysLogger = require('ain2');
    var console = new SysLogger();
    
    console.log('notice: %d', Date.now());
    console.info('info');
    console.error('error');
    
After launch in `/var/log/user` you can see the following:

    Dec  5 06:45:26 localhost ex.js[6041]: notice: 1291513526013
    Dec  5 06:45:26 localhost ex.js[6041]: info
    Dec  5 06:45:26 localhost ex.js[6041]: error

## Singleton logger

If you want to have a singleton that points to the same object whenever you do a require, use the following:

	require('ain2').getInstance();
	
If you use this, please be beware of this:

	require('ain2').getInstance() ===  require('ain2').getInstance();
	=> true
	
As opposed to:

	var SysLogger = require('ain2');
	new SysLogger() === new SysLogger();
	=> false
    
## Changing destinations

By default *ain* sets following destinations:

* `TAG` - `__filename`
* `Facility` - user (1)
* `HOSTNAME` - localhost
* `PORT` - 514
* `Transport` - UDP or Unix socket

You can change them by passing in the params to the constructor or by
using the `set` function. The `set` function is chainable.

    var SysLogger = require('ain2');
    var logger = new SysLogger({tag: 'node-test-app', facility: 'daemon', hostname: 'devhost', port: 3000});

    logger.warn('some warning');
    
... and in `/var/log/daemon.log`:

    Dec  5 07:08:58 devhost node-test-app[10045]: some warning
    
The `set` function takes one argument, a configuration object which can contain the following keys:
 * tag - defaults to __filename
 * facility - defaults to user
 * hostname - defaults to require('os').hostname()
 * port - defaults to 514
 * transport - defaults to 'UDP', can also be 'file'
 * messageComposer - a custom function to compose syslog messages

All of these are optional. If you provide a `hostname` transport is automatically set to UDP

`tag` and `hostname` arguments is just *RFC 3164* `TAG` and `HOSTNAME` of 
your messages.

`facility` is little more than just name. Refer to *Section 4.1.1* of 
[RFC 3164](http://www.faqs.org/rfcs/rfc3164.html) it can be:

    ##  String  Description
    -----------------------
     0  kern    kernel messages
     1  user    user-level messages
     2  mail    mail system
     3  daemon  system daemons
     4  auth    security/authorization messages
     5  syslog  messages generated internally by syslog daemon
     6  lpr     line printer subsystem
     7  news    network news subsystem
     8  uucp    UUCP subsystem
    16  local0  local use 0
    17  local1  local use 1
    18  local2  local use 2
    19  local3  local use 3
    20  local4  local use 4
    21  local5  local use 5
    22  local6  local use 6
    23  local7  local use 7

You can set the `facility` by `String` or `Number`:

    logger.set({tag: 'node-test-app', facility: 3});
    logger.set({tag: 'node-test-app', facility: 'daemon'});
    
Also you can set `TAG`, `Facility`, `HOSTNAME`, `PORT`, and `transport` separately by `setTag`, 
`setFacility`, `setHostname`, `setPort`, `setTransport` and `setMessageComposer` functions. All of them are chainable too.

You can get all destinations by these properties:

* `tag` TAG
* `facility` Numerical representation of RFC 3164 facility
* `hostname` HOSTNAME
* `port` PORT

## Custom message composer

    var SysLogger = require('ain2');
    var console = new SysLogger();

    console.setMessageComposer(function(message, severity){
        return new Buffer('<' + (this.facility * 8 + severity) + '>' +
                this.getDate() + ' ' + '[' + process.pid + ']:' + message);
    });

    
    //The default implementation looks this:


    SysLogger.prototype.composeSyslogMessage = function(message, severity) {
        return new Buffer('<' + (this.facility * 8 + severity) + '>' +
                this.getDate() + ' ' + this.hostname + ' ' + 
                this.tag + '[' + process.pid + ']:' + message);
    }

## Logging

As noticed before *ain* implements all `console` functions. Severity level is 
referenced to [RFC 3164](http://www.faqs.org/rfcs/rfc3164.html):

    #  String   Description
    -----------------------
    0  emerg    Emergency: system is unusable
    1  alert    Alert: action must be taken immediately
    2  crit     Critical: critical conditions
    3  err      Error: error conditions
    4  warn     Warning: warning conditions
    5  notice   Notice: normal but significant condition
    6  info     Informational: informational messages
    7  debug    Debug: debug-level messages

*Ain* `console`-like functions behaviour is fully compatible to *node.js* and 
logs messages with different severity levels: 

* `log` - notice (5)
* `info` - info (6)
* `warn` - warn (4)
* `error` - err (3)
* `dir` - notice (5)
* `time`, `timeEnd` - notice (5)
* `trace` - err (3)
* `assert` - err (3)

To log a message with the desired severity level you can use the `send` function:

    logger.send('message', 'alert');
    
The `send` function takes two arguments: message and optional severity level. By 
default, the severity level is *notice*.


CHANGELOG.md
# CHANGELOG

## v3.0.2

* Upgrade unix-dgram to 2.0.4
* Upgrade mocha to 8.3.2
* Replace deprecated [DEP0005] Buffer() to Buffer.from()

## v2.0.0

* Upgrade unix-dgram to 0.2.3 (josephfrazier)

## v1.5.2

* Fix socket bind problem in combination with cluster.js for node verions > v0.10.x (Issue #33)
* Have travis test more node versions

## v1.5.1

* Remove debug console.log (thx chazomaticus)

## v1.5.0

* Bubble up callbacks from udp and unix socket up to the caller

## v1.4.0

* Workaround for issue with node cluster on node 0.11.x (greenboxal)

## v1.3.1

* Added support for unix-dgram for node versions > 0.5
* Added travis build file to drive tests

## v1.2.1

* Reusing socket to send messages over UDP (Carlos Brito Lage/cblage)

## v1.1.1

Bugfix release

* The logging to file function crashed on unsupported systems due to
  immediate execution



## v1.1.0

* Custom message composer

## v1.0.0

WARNING: This upgrade is not API compatible to previous version.

* Change the API to actually use JavaScript's new operator to create loggers

		var SysLogger = require('ain2');
		var logger = new SysLogger({ port : 514, tag : 'myTag' });

* If you want to have singleton logger, use

		var logger = require('ain2').getInstance();

## v0.2.1

* Support for node v0.6.0 (Yoji Shidara/darashi)

## v0.2.0

* Support for unix sockets, for the 0.4.x branch of node (Parham Michael
  Ossareh/ossareh)

## v0.0.3

* Explicitly fall back to original `console` object to log send failures (Mark Wubben/novemberborn)
* Default hostname to machine name rather than localhost (Mark Wubben/novemberborn)
* Fixes to make jslint happy (Mark Wubben/novemberborn, Patrick
  Huesler/phuesler)
* Test server for local testing

## v0.0.2

* add support for custom host and port


test/ain.spec.js
var should = require('chai').should();
var ain = require('../index');
var fs = require('fs');

describe('ain', function(){
    it('should be a constructor', function(){
        ain.should.be.a('function');
    });
    it('allows fetching of a singleton', function(){
        ain.getInstance().should.be.equal(ain.getInstance());
    });
});


test/udp.spec.js
var dgram = require('dgram');
var should = require('chai').should();
var ain = require('../index');

describe('udp', function(){
    var server;
    var callback;
    var logger = new ain({port: 5514});
    before(function(){
        server = dgram.createSocket('udp4', function(msg, rinfo){
            if (callback) callback(msg);
        });
        server.bind(5514);
    });
    after(function(){
        server.close();
    });
    it('can send a udp buffer', function(done){
        logger.log('I will buff you up');
        callback = function(msg){
            msg.should.be.an.instanceof(Buffer);
            done();
        }
    });
});


test/unix.spec.js
var unixDgram = require('unix-dgram');
var should = require('chai').should();
var ain = require('../index');
var fs = require('fs');

describe('unix dgram', function(){
    var server;
    var callback;
    var logger = new ain({address: __dirname + '/socket', transport: 'unix_dgram'});
    before(function(done){
        server = unixDgram.createSocket('unix_dgram', function(msg, rinfo){
            if (callback) callback(msg);
        });
        server.bind(__dirname + '/socket');
        setTimeout(done, 1000); // give the socket time to open
    });
    after(function(){
        server.close();
        try{fs.unlinkSync(__dirname + '/socket'); } catch(e){};
    });
    it('can send a unix dgram buffer', function(done){
        logger.log('hello world');
        callback = function(msg){
            msg.should.be.an.instanceof(Buffer);
            msg.toString().should.contain('hello world');
            done();
        }
    });
});


index.js
var dgram = require('dgram');
var Buffer = require('buffer').Buffer;
var nodeConsole = console;

var DefaultHostname = require("os").hostname();
var DefaultAddress = "127.0.0.1";
var SingletonInstance = null;

var socket
var socketUsers = 0
var releaseTimeout
// HACK: On Node v0.11.x and v0.12.x bind the socket to ensure that it works when clustering
// This should be a temporary fix until node fixes the cluster module
var socketBindRequired = (function(){
    if(/^v0\.([0-9]|10)\..*$/.test(process.version)) {
      return false;
    }
    else {
      return true;
    }
})();

var socketErrorHandler = function (err) {

    if (err) {

        nodeConsole.error('socket error: ' + err)

    } else {

        nodeConsole.error('unknown socket error!')

    }



    if (socket !== undefined) {

        socket.close()

        socket = undefined

        socketUsers = 0

    }

}
var getSocket = function (type) {

    if (undefined === socket) {

        socket = dgram.createSocket(type);

        socket.on('error', socketErrorHandler)

        if (socketBindRequired && type == 'udp4') {
          socket.bind({
            port: 0,
            exclusive: true
          })
        }

    }

    ++socketUsers

    return socket

}
var releaseSocket = function () {

    --socketUsers

    if (0 == socketUsers && undefined === releaseTimeout) {

        releaseTimeout = setTimeout(function () {

            if (0 == socketUsers && socket !== undefined) {

                socket.close()

                socket = undefined

            }

            releaseTimeout = undefined

        }, 1000)

    }

}


var Transport = {
    UDP: function(message, severity, callback) {
        var self = this;
        var syslogMessage = this.composerFunction(message, severity);
        getSocket('udp4').send(syslogMessage,
                         0,
                         syslogMessage.length,
                         this.port,
                         this.address,
                         function(err, bytes) {
                             if(callback){
                                 callback(err, bytes);
                             }
                             self._logError(err, bytes);
                             releaseSocket();
                         }
                        );
    },
    unix_dgram: function(message, severity, callback){
        var self = this;
        var preambleBuffer = self.composerFunction('', severity);
        var formattedMessageBuffer = Buffer.isBuffer(message) ? message : new Buffer.from(message);
        var chunkSize = 2000 - preambleBuffer.length - 200;
        var numChunks = Math.ceil(formattedMessageBuffer.length / chunkSize);

        var fragments = [preambleBuffer];
        if (numChunks > 1){
            for (var i = 0; i < numChunks; i++){
                fragments.push(formattedMessageBuffer.slice(i * chunkSize, Math.min(formattedMessageBuffer.length, (i + 1) * chunkSize)),
                              new Buffer.from(' [' + (i + 1) + '/' + numChunks + ']', 'ascii'));
            }
        } else{
            fragments.push(formattedMessageBuffer);
        }

        var chunk = Buffer.concat(fragments);
        var socket = getSocket('unix_dgram');
        socket.send(chunk,
                    0,
                    chunk.length,
                    this.path,
                    function(err, bytes){
                        if(callback){
                          callback(err, bytes);
                        }
                        self._logError(err, bytes);
                        releaseSocket();
                    }
        );
    }
};

var Facility = {
    kern:   0,
    user:   1,
    mail:   2,
    daemon: 3,
    auth:   4,
    syslog: 5,
    lpr:    6,
    news:   7,
    uucp:   8,
    local0: 16,
    local1: 17,
    local2: 18,
    local3: 19,
    local4: 20,
    local5: 21,
    local6: 22,
    local7: 23
};

var Severity = {
    emerg:  0,
    alert:  1,
    crit:   2,
    err:    3,
    warn:   4,
    notice: 5,
    info:   6,
    debug:  7
};

// Format RegExp
var formatRegExp = /%[sdj]/g;

/**
 * Internal message formatting method
 * @param args [format, arg1, arg2, ...]
 * @returns
 */
function format(args) {
    var   util = require('util'),
    i    = 0,
    f    = args[0];

    // inspect the object if no format string was given as the first argument
    if (typeof f !== 'string') {
        var objects = [];
        for (i = 0; i < args.length; i++) {
            objects.push(util.inspect(args[i]));
        }
        return objects.join(' ');
    }


    i = 1;
    var str = String(f).replace(formatRegExp, function(x) {
        switch (x) {
            case '%s': return String(args[i++]);
            case '%d': return Number(args[i++]);
            case '%j': return JSON.stringify(args[i++]);
            default:
                return x;
        }
    });
    for (var len = args.length, x = args[i]; i < len; x = args[++i]) {
        if (x === null || typeof x !== 'object') {
            str += ' ' + x;
        } else {
            str += ' ' + util.inspect(x);
        }
    }
    return str;
}

/**
 * Syslog logger
 * @constructor
 * @returns {SysLogger}
 */
function SysLogger(config) {
    this._times = {};
    this._logError = function(err, other) {
        if(err){
            nodeConsole.error('Cannot log message via %s:%d', this.hostname, this.port);
        }
    }.bind(this);
    this.set(config);
    return this;
}

/**
 * Get singleton instance of SysLogger.
 * @returns {SysLogger}
 */
SysLogger.getInstance = function() {
    if(!SingletonInstance){
        SingletonInstance = new SysLogger();
    }
    return SingletonInstance;
};

/**
 * Init function, takes a configuration object. If a hostname is provided the transport is assumed
 * to be Transport.UDP
 * @param {Object} configuration object with the following keys:
 *          - tag       - {String}                  By default is __filename
 *          - facility  - {Facility|Number|String}  By default is "user"
 *          - hostname  - {String}                  By default is require("os").hostname()
 *          - address   - {String}                  By default is 127.0.0.1
 *          - port      - {Number}                  Defaults to 514
 *          - transport - {Transport|String}        Defaults to Transport.UDP
 */
SysLogger.prototype.set = function(config) {
    config = config || {} ;

    this.setTag(config.tag);
    this.setFacility(config.facility);
    this.setHostname(config.hostname);
    this.setAddress(config.address);
    this.setPort(config.port);
    this.setPath(config.path);
    this.setMessageComposer(config.messageComposer);
    this.setTransport(Transport.UDP);

    return this;
};

SysLogger.prototype.setTransport = function(transport) {
    this.transport = transport || Transport.UDP;
    if (typeof this.transport === 'string') {
        this.transport = Transport[this.transport] ;
    }
    if (typeof this.path === 'string' && this.path.length > 0){
        this.transport = Transport.unix_dgram
    }
    return this;
};

SysLogger.prototype.setTag = function(tag) {
    this.tag = tag || __filename;
    return this;
};

SysLogger.prototype.setFacility = function(facility) {
    this.facility = facility || Facility.user;
    if (typeof this.facility == 'string'){
        this.facility = Facility[this.facility];
    }
    return this;
};

SysLogger.prototype.setAddress = function(address) {
    this.address = address || DefaultAddress;
    return this;
};

SysLogger.prototype.setHostname = function(hostname) {
    this.hostname = hostname || DefaultHostname;
    return this;
};

SysLogger.prototype.setPort = function(port) {
    this.port = port || 514;
    return this;
};

SysLogger.prototype.setPath = function(path) {
    this.path = path || '';

    if (typeof this.path === 'string' && this.path.length > 0){
        try {
            dgram = require('unix-dgram');
        } catch(err){
            throw new Error('unix-dgram module not installed, cannot specify a unix socket path');
        }
    }

    return this;
};

SysLogger.prototype.setMessageComposer = function(composerFunction){
    this.composerFunction = composerFunction || this.composeSyslogMessage;
    return this;
};

/**
 * Send message
 * @param {String} message
 * @param {Severity} severity
 */
SysLogger.prototype._send = function(message, severity, callback) {
    this.transport(message, severity, callback) ;
};

/**
 * Send formatted message to syslog
 * @param {String} message
 * @param {Number|String} severity
 */
SysLogger.prototype.send = function(message, severity, callback) {
    if (typeof severity == 'string'){
        severity = Severity[severity];
    }
    this._send(message, severity, callback);
};

SysLogger.prototype.formatAndSend = function(argumentsObject, severity) {
    var args          = Array.prototype.slice.call(argumentsObject);
    var callback      = undefined;
    var message       = undefined;


    if(Object.prototype.toString.call(args[args.length - 1]) == "[object Function]"){
      callback = args.pop();
    }
    this._send(format(args), severity, callback);
};




/**
 * Send log message with notice severity.
 */
SysLogger.prototype.log = function() {
    this.formatAndSend(arguments, Severity.notice);
};
/**
 * Send log message with info severity.
 */
SysLogger.prototype.info = function() {
    this.formatAndSend(arguments, Severity.info);
};
/**
 * Send log message with warn severity.
 */
SysLogger.prototype.warn = function() {
    this.formatAndSend(arguments, Severity.warn);
};
/**
 * Send log message with err severity.
 */
SysLogger.prototype.error = function() {
    this.formatAndSend(arguments, Severity.err);
};
/**
 * Send log message with debug severity.
 */
SysLogger.prototype.debug = function() {
    this.formatAndSend(arguments, Severity.debug);
};

/**
 * Log object with `util.inspect` with notice severity
 */
SysLogger.prototype.dir = function(object, callback) {
    var util = require('util');
    this._send(util.inspect(object) + '\n', Severity.notice);
};

SysLogger.prototype.trace = function(label, callback) {
    var err = new Error();
    err.name = 'Trace';
    err.message = label || '';
    Error.captureStackTrace(err, arguments.callee);
    this.error(err.stack, callback);
};

SysLogger.prototype.assert = function(expression) {
    if (!expression) {
        var arr = Array.prototype.slice.call(arguments, 1);
        this._send(format(arr), Severity.err);
    }
};


/**
 * Compose syslog message
 */
SysLogger.prototype.composeSyslogMessage = function(message, severity) {
    return new Buffer.from('<' + (this.facility * 8 + severity) + '>' +
                      this.getDate() + ' ' + this.hostname + ' ' +
                      this.tag + '[' + process.pid + ']:' + message);
}

SysLogger.prototype.time = function(label) {
    this._times[label] = Date.now();
};
SysLogger.prototype.timeEnd = function(label) {
    var duration = Date.now() - this._times[label];
    this.log('%s: %dms', label, duration);
};

/**
 * Get current date in syslog format.
 * @returns {String}
 */
SysLogger.prototype.getDate = function() {
    return new Date().toISOString();
}

SysLogger.prototype.leadZero = function(n) {
    if (n < 10) {
        return '0' + n;
    } else {
        return n;
    }
}

module.exports = SysLogger;
